algorithms/src/snark/varuna/ahp/matrices.rs

```diff
@@ -37,19 +36,25 @@ use rayon::prelude::*;
 // This function converts a matrix output by Zexe's constraint infrastructure
 // to the one used in this crate.
 pub(crate) fn to_matrix_helper<F: Field>(
-    matrix: &[Vec<(F, VarIndex)>],
+    matrix: Vec<Vec<(F, VarIndex)>>,
     num_input_variables: usize,
 ) -> Result<Matrix<F>> {
-    cfg_iter!(matrix)
+    cfg_into_iter!(matrix)
         .map(|row| {
-            let mut row_map = BTreeMap::new();
-            for (val, column) in row.iter() {
-                ensure!(*val != F::zero(), "matrix entries should be non-zero");
+            let mut row_map = Vec::with_capacity(row.len());
+            for (val, column) in row {
+                ensure!(val != F::zero(), "matrix entries should be non-zero");
                 let column = match column {
-                    VarIndex::Public(i) => *i,
+                    VarIndex::Public(i) => i,
                     VarIndex::Private(i) => num_input_variables + i,
                 };
-                *row_map.entry(column).or_insert_with(F::zero) += *val;
+                match row_map.binary_search_by_key(&column, |(c, _)| *c) {
+                    Ok(idx) => row_map[idx].1 += val,
+                    Err(idx) => {
+                        row_map.insert(idx, (column, F::zero()));
+                        row_map[idx].1 += val;
```

Comment:
```suggestion
                        row_map.insert(idx, (column, val));
```
question: could we do this instead?

Reply:
Yes that would be equivalent.

Reply:
sweet, because it should be a bit faster 🔥